app/scripts/lib/router.js

```diff
@@ -159,17 +159,43 @@ define(function (require, exports, module) {
     },
 
     /**
-     * Redirect the user to the best suitable OAuth flow
+     * Redirect the user to the best suitable OAuth flow.
      */
     redirectToBestOAuthChoice: function () {
-      var account = this.user.getChooserAccount();
+      var self = this;
       var route = '/oauth/signin';
 
-      if (account.isDefault()) {
-        route = '/oauth/signup';
-      }
+      var account = this.user.getChooserAccount();
 
-      return this.navigate(route, { replace: true, trigger: true });
+      // Attempt to get email address from relier
+      var email = this.broker.relier.get('email');
+
+      // Check whether email address corresponds to an existing account and
+      // navigate to correct login/signup page.
+      if (email) {
+        return account.checkAccountExistsByEmail(email)
+        .then(function (exists) {
+          if (! exists) {
+            route = '/oauth/signup';
+          }
+        }, function (err) {
+          // In case of error, choose navigation page based on
+          // whether account is a default account
+          if (account.isDefault()) {
+            route = '/oauth/signup';
+          }
+        })
+        .then(function () {
+          return self.navigate(route, { replace: true, trigger: true });
+        });
+      } else {
+        // If no email in relier, choose navigation page based on
+        // whether account is a default account
+        if (account.isDefault()) {
+          route = '/oauth/signup';
+        }
+        return self.navigate(route, { replace: true, trigger: true });
```

Comment:
I feel like there is a way to linearize this function so there is only one call to `self.navigate`, but I'm too tired to think of it.